File: Day8_Subarrays_With_K_Distinct_Integers/SubarraysWithKDistinct.py
class Solution(object):
    def numberOfStringsLessOrEqualThreeChar(self,nums,k):
        d=dict()
        p1=0
        p2=0
        ans=0
        while p1<len(nums):
            if nums[p1] in d:
                d[nums[p1]]+=1
            else:
                d[nums[p1]]=1
            if len(d)<=k:
                ans+=(p1-p2+1)
            else:
                while len(d)>k:
                    d[nums[p2]]-=1
                    if d[nums[p2]]==0:
                        del d[nums[p2]]
                    p2+=1
                ans+=(p1-p2+1)
            p1+=1
        return ans

    
    def subarraysWithKDistinct(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        return self.numberOfStringsLessOrEqualThreeChar(nums,k)-self.numberOfStringsLessOrEqualThreeChar(nums,k-1)
        

File: Day8_Subarrays_With_K_Distinct_Integers/SubarraysWithKDistinct.java
class Solution {
    public int subarraysWithLessOrKDistinct(int[] nums,int k){
        int p1=0;
        int p2=0;
        HashMap<Integer,Integer> m=new HashMap<>();
        int ans=0;
        while(p1<nums.length){

            if(m.containsKey(nums[p1])){
                m.put(nums[p1],m.get(nums[p1])+1);
            }
            else{
                m.put(nums[p1],1);
            }
            if( m.size()<=k){
                ans+=(p1-p2+1);
            }
            else{
                while(m.size()>k){
                    m.put(nums[p2],m.get(nums[p2])-1);
                    if (m.get(nums[p2])==0){
                        m.remove(nums[p2]);
                    }
                    p2+=1;
                }
                ans+=(p1-p2+1);
            }
            p1+=1;

        }
        return ans;
    }
    public int subarraysWithKDistinct(int[] nums, int k) {
        return subarraysWithLessOrKDistinct(nums,k)-subarraysWithLessOrKDistinct(nums,k-1);
        
    }
}
